torch/_subclasses/fake_tensor.py

```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+        elif isinstance(arg, torch.Tensor):
+            raise _BypassDispatchCache("non-fake tensor")
+        else:
+            # It's important to capture the type of the arg since, e.g., 1 and 1.0
+            # hash to the same value, but can produce different dtypes for the
+            # output tensor.
+            return (type(arg), arg)
+
+    def _create_cache_entry(
+        self, key: _DispatchCacheKey, args: Tuple[Any], output: FakeTensor
+    ) -> _DispatchCacheEntry:
+        """
+        Create an entry in the cache for the given 'output' Tensor.
+        """
+        # Some ops return tuples of Tensors, but it's rare, so avoid
+        # the complexity of caching other types.
+        if not isinstance(output, FakeTensor):
+            raise _BypassDispatchCache("non-FakeTensor output")
+
+        # Avoid caching FakeTensors with constants attached since those
+        # can be invalidated.
+        if output.constant is not None:
+            raise _BypassDispatchCache("constant attribute")
+
+        # TODO: support caching sparse outputs?
+        if output.is_sparse:
+            raise _BypassDispatchCache("sparse output")
+
+        # If this is an in-place op, we just record which input arg is aliased.
+        entry = None
+        for idx in range(len(args)):
+            if id(args[idx]) == id(output):
+                entry = _DispatchCacheEntry(None, idx)
+                break
+
+        # Otherwise, record the output tensor's metadata.
+        if not entry:
+            entry = _DispatchCacheEntry(extract_tensor_metadata(output), None)
+
+        FakeTensorMode.cache[key] = entry
+
+    def _output_from_cache_entry(
+        self, entry: _DispatchCacheEntry, func: OpOverload, args: Tuple[Any]
+    ) -> FakeTensor:
+        """
+        Create a new FakeTensor from the cache entry.
+        """
+        if entry.arg_idx is not None:
+            # This is an in-place op; return the aliased arg.
+            return args[entry.arg_idx]
+
+        # Create a new empty FakeTensor with the cached metadata.
+        metadata = entry.metadata
+        assert not metadata.is_quantized
+        assert not metadata.is_sparse
+
+        with torch.inference_mode(metadata.is_inference):
+            empty = torch.empty_strided(
+                metadata.shape,
+                metadata.stride,
+                dtype=metadata.dtype,
+                layout=metadata.layout,
+                device="meta",
+                requires_grad=metadata.requires_grad,
+            )
+
+        torch._C._set_conj(empty, metadata.is_conj)
+        torch._C._set_neg(empty, metadata.is_neg)
+
+        if func.is_view or metadata.storage_offset != 0:
+            if func.is_view:
+                # For view ops, the storage should be the same as the input.
+                tensor_args = [t for t in args if isinstance(t, torch.Tensor)]
+                assert len(tensor_args) == 1
+                storage = tensor_args[0].untyped_storage()
```

Comment:
Hmm, we do we have this invariant? That if it's a view there is only one tensor and it's aliased. For instance, `index_put_` does not follow this. 

What if `index_put_ is called with `self` as a kwarg ? Seems like we need to handle kwargs here as well.

Reply:
> For instance, index_put_ does not follow this.

I think this is actually true in practice today for ops that are `.is_view` (index_put_ doesn't register as a view op, since it mutates its input)
```
>>> torch.ops.aten.index_put_.default.is_view
False
```

I am a bit worried that there are edge cases we are missing for view support, but I can't think of any. Since e.g. a lot of the complexity for properly returning a view is around getting the autograd metadata correct (the `.grad_fn` and `.is_view` fields), and we don't have to worry about that here since we're below autograd

Reply:
My takeaway (but elias can tell me if he disagrees) is:

(1) the code as written assumes that if `op.is_view` is true, then it must have a **single** tensor arg, guaranteed to be `args[0]`, which is the tensor arg that aliases the output tensor

(2) `is_view` only returns true for ops with "immutable aliasing", aka it will not return true for inplace ops like `.index_put_`, or `add_.Tensor`

(3) Therefore I think this is a safe assumption to make today - that assumption is the case for all view ops today that are a part of ATen